### ISSUE TITLE
BlobEncode firebird

### ISSUE THREAD

**Issue by sund1** [ADOdb/ADOdb#743]
in firebird driver BlobEncode uses $ this -> _ connectionID
with an active transaction, the request crashes with the error "0: invalid BLOB ID"

do I need to add a check to function BlobEncode ($ blob) that there is a transaction or am I doing something wrong?
```
$db->Execute('insert into table (test) values (?)', $db->BlobEncode('test'));
```
```
	function BlobEncode( $blob )
	{
	   if ($this->_transactionID) { // add?
               $blobid = ibase_blob_create( $this->_transactionID); // add?
            } else {
               $blobid = ibase_blob_create( $this->_connectionID);
            }
            ibase_blob_add( $blobid, $blob );
	    return ibase_blob_close( $blobid );
	}
	```

**Comment by mnewnham** [comment]
Just to start this issue to let you know that the firebird driver has been completely re-written and is available by grabbing the master code. Obviously because there have been so many major changes, I'd like to spend as little time as possible updating the 5.21 release on this. Would it be possible for you to just check out this new version to see if it resolves your problem?

**Comment by sund1** [comment]
> Just to start this issue to let you know that the firebird driver has been completely re-written and is available by grabbing the master code. Obviously because there have been so many major changes, I'd like to spend as little time as possible updating the 5.21 release on this. Would it be possible for you to just check out this new version to see if it resolves your problem?

with blobencode the problem is solved
in other blob* functions you also need to select connectionID or transactiondID


another mistake
```
$db->Execute(query1);
$db->Execute(query2);

$db->BeginTrans();
$db->Execute(query3);
$db->CommitTrans();

$db->GetOne('select list(key) from table') // = null and 
Warning: fbird_blob_info(): invalid BLOB ID in /var/www/html/web_git/vendor/adodb/adodb-php/drivers/adodb-firebird.inc.php on line 579
Warning: fbird_blob_open(): invalid BLOB ID in /var/www/html/web_git/vendor/adodb/adodb-php/drivers/adodb-firebird.inc.php on line 580
Warning: fbird_blob_get() expects parameter 1 to be resource, bool given in /var/www/html/web_git/vendor/adodb/adodb-php/drivers/adodb-firebird.inc.php on line 589
Warning: fbird_blob_close() expects parameter 1 to be resource, bool given in /var/www/html/web_git/vendor/adodb/adodb-php/drivers/adodb-firebird.inc.php on line 592
```

there are no errors on version 5.20.18


also:
```
$db = ADONewConnection(firebird) 

$db->Execute('update table1 set field1 = 123');// #1


$db->BeginTrans();
$db->Execute('update table1 set field1 = 555'); // return deadlock because #1 not not committed (in the old version, the request will be executed)
$db->CommitTrans();
```

**Comment by mnewnham** [comment]
Addressing your notes individually as I have time: 
My PHP is 7.4.13
My Platform is Windows
My firebird driver is the new driver from [here](https://github.com/FirebirdSQL/php-firebird/releases/tag/v1.1.1)

**Blob functions**
>in other blob* functions you also need to select connectionID or transactiondID

Could you provide some code snippets please?

**Deadlock**
I have tested your program with ADOdb version 5.20.19 and master, and I get the same result for both: A deadlock error. To me, this would be the expected result as the transaction handling functions have not changed at all between the 2 versions, I guess I would need to understand deeper how you are able to avoid the deadlock in the earlier version.

**Comment by sund1** [comment]
php 7.2.34

```
   $db = ADONewConnection($driver);
   $db->charSet = $charset;
   $db->dialect = 3;
   $db->fmtDate = 'd.m.Y';
   $db->fmtTimeStamp = 'd.m.Y H:i:s';

   $db->ibase_datefmt = '%d.%m.%Y';
   $db->ibase_timestampfmt = "%d.%m.%Y %H:%M:%S";
   $db->ibase_timefmt = "%H:%M:%S";

   $db->Connect($host, $user, $pass, $name);   
   
   $rs = $db->Execute('update TEST set t_name = t_name where t_prefix = 22567');

   $db->BeginTrans();
   $rs = $db->Execute('update TEST set t_name = t_name where t_prefix = 22567');
   $db->CommitTrans(); 
  
  
  // adodb 5.20.19 all ok
  // adodb 5.21.0 update TEST set t_name = t_name where t_prefix = 22567   0: deadlock update conflicts with concurrent update concurrent transaction number is 10670390 

   
   ```
   
   
   ```
   driver ADODB_firebird 5.21 
   trans_id != connection_id
   ```
![image](https://user-images.githubusercontent.com/33967850/128127154-1974f624-0705-4ba8-b3d0-2886b3126df2.png)

     ```
     driver ADODB_firebird 5.20.19 
     ```
![image](https://user-images.githubusercontent.com/33967850/128127341-1903d9e7-4ef2-4a17-b115-742e1fe62872.png)
 ```
 adodb_ibase -> BeginTrans()
 connection_id = transaction_id
 ```
 
![image](https://user-images.githubusercontent.com/33967850/128127381-db416561-ef33-4638-a30c-419825191435.png)

```
Blob functions
i only tested blobencode

the other needs to be redone probably like this
```
![image](https://user-images.githubusercontent.com/33967850/128128096-200bbc12-8c9a-4c13-ae16-686e78a24ed0.png)
-> 
![image](https://user-images.githubusercontent.com/33967850/128128173-731050ad-c066-452c-a18f-8b6640877144.png)

**Comment by mnewnham** [comment]
It appears that you are actually using the ibase driver in 5.20.19. Is that correct?  Using your program above I can show that by simply swapping the adodb driver from firebird to ibase, the deadlock issue no longer occurs.

**Comment by dregad** [comment]
@mnewnham wrote in https://github.com/ADOdb/ADOdb/issues/743#issuecomment-890270264
> the firebird driver has been completely re-written and is available by grabbing the master code

@sund1 There might have been a bit of confusion here, because the updated FireBird driver was in fact *not* available in _master_ at the time Mark wrote this - my apologies for that.

I have now merged the pending pull request (see c823ab67), so could you please test again with the latest code from the _master_ branch and let us know how it goes ?

**Comment by dregad** [comment]
@sund1 without feedback from you for a month, I assume that this is no longer an issue for you and will close this issue. Feel free to reopen should the problem persists.
